ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedColumnReader.java

```diff
@@ -0,0 +1,577 @@
+    this.pageReader = pageReader;
+    this.maxDefLevel = descriptor.getMaxDefinitionLevel();
+    this.skipTimestampConversion = skipTimestampConversion;
+
+    DictionaryPage dictionaryPage = pageReader.readDictionaryPage();
+    if (dictionaryPage != null) {
+      try {
+        this.dictionary = dictionaryPage.getEncoding().initDictionary(descriptor, dictionaryPage);
+        this.isCurrentPageDictionaryEncoded = true;
+      } catch (IOException e) {
+        throw new IOException("could not decode the dictionary for " + descriptor, e);
+      }
+    } else {
+      this.dictionary = null;
+      this.isCurrentPageDictionaryEncoded = false;
+    }
+  }
+
+  void readBatch(
+    int total,
+    ColumnVector column,
+    TypeInfo columnType) throws IOException {
+
+    int rowId = 0;
+    while (total > 0) {
+      // Compute the number of values we want to read in this page.
+      consume();
+      int leftInPage = (int) (endOfPageValueCount - valuesRead);
+      int num = Math.min(total, leftInPage);
+      if (isCurrentPageDictionaryEncoded) {
+        LongColumnVector dictionaryIds = new LongColumnVector();
+        // Read and decode dictionary ids.
+        readDictionaryIDs(num, dictionaryIds, rowId);
+        decodeDictionaryIds(rowId, num, column, dictionaryIds);
+      } else {
+        // assign values in vector
+        PrimitiveTypeInfo primitiveColumnType = (PrimitiveTypeInfo) columnType;
+        switch (primitiveColumnType.getPrimitiveCategory()) {
+        case INT:
+        case BYTE:
+        case SHORT:
+          readIntegers(num, (LongColumnVector) column, rowId);
+          break;
+        case DATE:
+        case INTERVAL_YEAR_MONTH:
+        case LONG:
+          readLongs(num, (LongColumnVector) column, rowId);
+          break;
+        case BOOLEAN:
+          readBooleans(num, (LongColumnVector) column, rowId);
+          break;
+        case DOUBLE:
+          readDoubles(num, (DoubleColumnVector) column, rowId);
+          break;
+        case BINARY:
+        case STRING:
+        case CHAR:
+        case VARCHAR:
+          readBinaries(num, (BytesColumnVector) column, rowId);
+          break;
+        case FLOAT:
+          readFloats(num, (DoubleColumnVector) column, rowId);
+          break;
+        case DECIMAL:
+          readDecimal(num, (DecimalColumnVector) column, rowId);
+        case INTERVAL_DAY_TIME:
+        case TIMESTAMP:
+        default:
+          throw new IOException("Unsupported");
+        }
+      }
+      rowId += num;
+      total -= num;
+    }
+  }
+
+  private void consume() throws IOException {
+    int leftInPage = (int) (endOfPageValueCount - valuesRead);
+    if (leftInPage == 0) {
+      readPage();
+    }
+  }
+
+  private void readDictionaryIDs(
+    int total,
+    LongColumnVector c,
+    int rowId) throws IOException {
+    int left = total;
+    while (left > 0) {
+      readRepetitionAndDefinitionLevels();
+      if (definitionLevel >= maxDefLevel) {
```

Comment:
It could be greater than maxDefLevel according to the definition of null in this [document](https://github.com/Parquet/parquet-mr/wiki/The-striping-and-assembly-algorithms-from-the-Dremel-paper). 
>When a field is not defined, the definition level will be smaller than the definition level of this field.